core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -295,7 +295,9 @@ private[spark] class ExecutorAllocationManager(
 
       // If the new target has not changed, avoid sending a message to the cluster manager
       if (numExecutorsTarget < oldNumExecutorsTarget) {
-        client.requestTotalExecutors(numExecutorsTarget)
+        val (localityAwarePendingTasks, preferredLocalities) = listener.executorPlacementHints()
```

Comment:
Can we cache the result of `executorPlacementHints` and only update it when stages come or go?